presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AbstractExpressionRewriteRuleSet.java

```diff
@@ -0,0 +1,234 @@
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.plan.AggregationNode;
+import com.facebook.presto.sql.planner.plan.ApplyNode;
+import com.facebook.presto.sql.planner.plan.Assignments;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.JoinNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.planner.plan.TableScanNode;
+import com.facebook.presto.sql.planner.plan.ValuesNode;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.google.common.collect.ImmutableSet;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+
+public abstract class AbstractExpressionRewriteRuleSet
+{
+    abstract Expression rewrite(Expression expression, Rule.Context context);
+
+    public Set<Rule> getRules()
+    {
+        return ImmutableSet.of(
+                new ProjectExpressionRewrite(),
+                new AggregationExpressionRewrite(),
+                new FilterExpressionRewrite(),
+                new TableScanExpressionRewrite(),
+                new JoinExpressionRewrite(),
+                new ValuesExpressionRewrite(),
+                new ApplyExpressionRewrite()
+        );
+    }
+
+    private static class Patterns
+    {
+        private static final Pattern PROJECT = Pattern.typeOf(ProjectNode.class);
+        private static final Pattern AGGREGATION = Pattern.typeOf(AggregationNode.class);
+        private static final Pattern FILTER = Pattern.typeOf(FilterNode.class);
+        private static final Pattern TABLE_SCAN = Pattern.typeOf(TableScanNode.class);
+        private static final Pattern JOIN = Pattern.typeOf(JoinNode.class);
+        private static final Pattern VALUES = Pattern.typeOf(ValuesNode.class);
+        private static final Pattern APPLY = Pattern.typeOf(ApplyNode.class);
+    }
+
+    final class ProjectExpressionRewrite
+            implements Rule
+    {
+        @Override
+        public Pattern getPattern()
+        {
+            return Patterns.PROJECT;
+        }
+
+        @Override
+        public Optional<PlanNode> apply(PlanNode node, Context context)
+        {
+            ProjectNode projectNode = (ProjectNode) node;
+            Assignments assignments = projectNode.getAssignments().rewrite(x -> AbstractExpressionRewriteRuleSet.this.rewrite(x, context));
+            return Optional.of(new ProjectNode(node.getId(), projectNode.getSource(), assignments));
```

Comment:
return `Optional.empty()` when there was no changes, here in all the other rules in this file